README.rst

```diff
@@ -3,56 +3,71 @@ openedx-completion-aggregator
+
+This Open edX plugin, ``openedx-completion-aggregator`` watches course activity and asynchronously updates database tables with "aggregate" completion data. "Aggregate" data means completion data summed up over all XBlocks into a course and aggregated at higher levels, like the subsection, section, and course level. The completion aggregator provides a REST API that can provide near-instant answers to queries such as:
+
+* What % complete are each of the courses that I'm enrolled in?
+* What % of each section in Course X have my students completed?
+* What is the average completion % among all enrolled students in a course?
+
+API Details
+-----------
+
+For details about how the completion aggregator's REST APIs can be used, please refer to `the docstrings in views.py <https://github.com/open-craft/openedx-completion-aggregator/blob/master/completion_aggregator/api/v1/views.py#L24>`_.
+
+Installation and Configuration
+------------------------------
+
+openedx-completion-aggregator uses the pluggable django app pattern to ease installation. To use in edx-platform, do the following:
+
+1.  Install the app into your virtualenv::
 
         $ pip install openedx-completion-aggregator
 
-2.  [Optional] You may override the set of registered aggregator block types in
-    your lms.env.json file::
+2.  [Optional] You may override the set of registered aggregator block types in your ``lms.yml`` file::
 
         ...
-        "COMPLETION_AGGREGATOR_BLOCK_TYPES": {
-            "course",
-            "chapter",
-            "subsection",
-            "vertical"
-        },
+        COMPLETION_AGGREGATOR_BLOCK_TYPES:
+            - course
+            - chapter
+            - subsection
+            - vertical
         ...
 
 
-3.  By default, completion is aggregated with each created or updated
-    BlockCompletion.  In most production instances, you will want to calculate
-    aggregations asynchronously.  To enable asynchronous calculation for your
-    installation, set the following in your lms.env.json file::
+3.  By default, completion is aggregated synchronously (with each created or updated BlockCompletion). While that is ideal for development, in most production instances, you will want to calculate aggregations asynchronously for better performance.  To enable asynchronous calculation for your installation, set the following in your ``lms.yml`` file::
 
         ...
-        "COMPLETION_AGGREGATOR_ASYNC_AGGREGATION": true,
+        COMPLETION_AGGREGATOR_ASYNC_AGGREGATION: true
         ...
 
-    Then configure up a pair of cron jobs to run `./manage.py
-    run_aggregator_service` and `./manage.py run_aggregator_cleanup` as often
+    Then configure up a pair of cron jobs to run ``./manage.py
```

Comment:
If the aggregator is installed on an existing instance, then it's sometimes desirable to fill "Aggregate" data for the existing courses. There is the [reaggregate_course](https://github.com/open-craft/openedx-completion-aggregator/blob/master/completion_aggregator/management/commands/reaggregate_course.py) management command, which prepares data that will be aggregated during the next `run_aggregator_service` run.
However that are a few points that might cause problems while doing this if you are running it for all courses on a big instance. People deploying this should be sure that they have:
1. Disabled cronjobs (if they have already enabled them). Otherwise these will clog Redis.
1. Enough computing power to run the aggregator for the first time. For our case we did a simulation by moving all data from prod to stage and measuring how long this will take. For one instance they were taking about 125h, so for the prod [we went with](https://tasks.opencraft.com/browse/BB-2296?focusedCommentId=148978&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-148978) disabling all existing Celery workers and spawning 3 new `m5.xlarge` instances running only workers with concurrency set to 4 (so 12 processes total).
1. Enough DB computing power. We have upgraded our RDS for this process, to be sure that the heavy parallelization won't make LMS unusable.
1. Enough DB storage (described [here](https://forum.opencraft.com/t/opencraft-ops-review/533/5)). "Aggregate" data is quite heavy - for us it added 20GB to the DB size during the initial run. We had RDS auto scaling enabled, but it throttled the instance significantly. Rescaling it manually fixed the issue, so it might be worth to scale DB upfront.

Reply:
Thanks, this is very valuable information and I've add this in to the README. Please take a look.

Is it `reaggregate_course` or `run_aggregator_service` that takes 125h? I assume the latter, but I thought you said you added a lock to prevent overlapping runs of it, so how did you use concurrent workers to speed it up?

Reply:
@bradenmacdonald, it's `run_aggregator_service`. `reaggregate_course` didn't need more than a few minutes.

> I thought you said you added a lock to prevent overlapping runs of it, so how did you use concurrent workers to speed it up?

Oh, I now remember the specific use case we had in mind here.
The reason we implemented this is because we have this specified in Ansible as:
```yaml
EDXAPP_ADDITIONAL_CRON_JOBS:
  - name: "run aggregator service"
    user: "{{ edxapp_user }}"
    job: ". {{ edxapp_app_dir }}/edxapp_env && {{ edxapp_venv_bin }}/python {{ edxapp_code_dir }}/manage.py lms run_aggregator_service --settings={{ edxapp_settings }} >/dev/null 2>&1"
    minute: "*/30"
    hour: "*"
    day: "*"
  - name: "run aggregator cleanup"
    user: "{{ edxapp_user }}"
    job: ". {{ edxapp_app_dir }}/edxapp_env && {{ edxapp_venv_bin }}/python {{ edxapp_code_dir }}/manage.py lms run_aggregator_cleanup --settings={{ edxapp_settings }} >/dev/null 2>&1"
    minute: "15"
    hour: "*"
    day: "*"
```
Therefore this job is present on each instance in the auto scaling group. This would fire all of them simultaneously, so we are using Redis cache locks here (which use atomic operations) to be sure that only one instance is able to run this.

The lock is only for scheduling Celery tasks, not for running them. We didn't create the locks to protect ourselves from the infinite loops, but this can be an issue if you run the `run_aggregator_service` too often while having tons of new block completions. This is not a very likely scenario, but it's technically possible. This could potentially be mitigated by marking `StaleCompletion`s as `resolved` once they are scheduled for the Celery task (or by adding a `scheduled` field to this model), but I'm not fully aware whether this would not cause any side effects.
Sorry for the confusion on this!

Reply:
@Agrendalath 
I see, thanks.

Ok, so does it happen sometimes that `run_aggregator_service` (which just enqueues the tasks, not actually processes them) can take longer than 30 minutes? I just want to ensure this wording is still correct:

> A lock is used to prevent overlapping runs of run_aggregator_service. If you have a lot of data on your instance such that a single run of run_aggregator_service takes longer than 30 minutes, you must configure COMPLETION_AGGREGATOR_AGGREGATION_LOCK_TIMEOUT_SECONDS to a larger value to prevent releasing the lock too early as a result of a timeout.

Reply:
@bradenmacdonald, good point, `run_aggregator_service` shouldn't take that much time in the regular usage (not counting its manual run after `reaggregate_course`). We can remove this part and add it in case we need some locks shared between celery tasks in the future.